src/lib/ilqp.js

```diff
@@ -0,0 +1,202 @@
+'use strict'
+
+const co = require('co')
+const Packet = require('../utils/packet')
+const debug = require('debug')('ilp:ilqp')
+const moment = require('moment')
+const BigNumber = require('bignumber.js')
+const uuid = require('uuid')
+const { safeConnect, startsWith, wait, xor, omitUndefined } =
+  require('../utils')
+
+const DEFAULT_MESSAGE_TIMEOUT = 5000
+const DEFAULT_EXPIRY_DURATION = 10
+
+function * _handleConnectorResponses (connectors, promises) {
+  if (connectors.length === 0) {
+    throw new Error('no connectors specified')
+  }
+
+  const quotes = []
+  const errors = []
+
+  for (let c = 0; c < connectors.length; ++c) {
+    try {
+      const quote = yield promises[c]
+      if (quote) quotes.push(quote)
+    } catch (err) {
+      errors.push(connectors[c] + ': ' + err.message)
+    }
+  }
+
+  if (quotes.length === 0) {
+    throw new Error('Errors occurred during quoting: ' +
+      errors.join(', '))
+  }
+
+  return quotes
+}
+
+function _sendAndReceiveMessage ({
+  plugin,
+  method,
+  message,
+  timeout
+}) {
+  const id = message.data.id = message.data.id || uuid()
+  debug('sending message:', JSON.stringify(message))
+
+  const responded = new Promise((resolve, reject) => {
+    function onIncomingMessage (response) {
```

Comment:
the function definition is so that it can refer to itself when removing the `incoming_message` listener